walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,103 @@
+      return timesForMeeting;
+    }
+ 
+    // Assume optional attendees can't be included until
+    // slot is found
+    boolean optionalsIncluded = false;
+
+    // Begin search at start of day
+    int beginSlot = TimeRange.START_OF_DAY;
+
+    // Keep track of time slots for optional attendees
+    int beginSlotOptional = beginSlot;
+
+    for (Event event : events) {
+      // Check for conflicts
+      boolean noConflict = checkForConflict(event, request.getAttendees());
+      boolean noOptionalConflict = checkForConflict(event, request.getOptionalAttendees());
+
+      // If optionals can be included do original search method
+      if (optionalsIncluded) {
+        // No conflicts, move on
+        if ( noConflict && noOptionalConflict) {
+          continue;
+        }
+
+        beginSlot = addSlotIfItFits(event, request, beginSlot, timesForMeeting);
+
+      } else {
+        // Check to see if optional attendees can be included
+        if (noOptionalConflict) {
+          if (slotFits(event, request, beginSlotOptional)) {
+            optionalsIncluded = true;
+          }
+        } else {
+          if (slotFits(event, request, beginSlotOptional)) {
+            optionalsIncluded = true;
+            beginSlot = addSlotIfItFits(event, request, beginSlotOptional, timesForMeeting);
+            continue;
+          } else {
+            beginSlotOptional = event.getWhen().end();
+          }
+        }
+
+        // No conflict move on
+        if (noConflict) {
+          continue;
+        }
+
+        beginSlot = addSlotIfItFits(event, request, beginSlot, timesForMeeting);
+      }
+    }
+    
+    // Check for open slot at end of day
+    if (beginSlot < TimeRange.END_OF_DAY) {
+      timesForMeeting.add(TimeRange.fromStartEnd(beginSlot, TimeRange.END_OF_DAY, true));
+    }
+    return timesForMeeting;
+  }
+
+  /* Checks if the set of an event's attendees does not intersect the
+   * set of the request's attendees
+   */
+  public boolean checkForConflict(Event event, Collection<String> attendees) {
+    return Collections.disjoint(event.getAttendees(), attendees);
+  }
+
+  public int addSlotIfItFits(Event event, MeetingRequest request, 
+                             int beginSlot, ArrayList<TimeRange> timesForMeeting) {
+    // On conflict end time slot before event
+    int endSlot = event.getWhen().start();
+
+    // Check that request fits in time slot and add time slot
+    if ( slotFits(event, request, beginSlot) ) {
+      timesForMeeting.add(TimeRange.fromStartEnd(beginSlot, endSlot, false));
+    }
+
+    // Start new slot after event ends
+    if (beginSlot < event.getWhen().end()) {
+      beginSlot = event.getWhen().end();
+    }
+
+    return beginSlot;
+  }
+
+  // Returns false if the current slot is too short for the request
+  public boolean slotFits(Event event, MeetingRequest request, int beginSlot) {
+    int endSlot = event.getWhen().start();
```

Comment:
I would recommend moving this into a variable that called "eventLength" or something, and doing the math up here for that. Than your result length is "eventLength >= duration" which is clearer than what you have now.

Reply:
done